From 445e67805dd2ca6c3a2363b2ec9e163c62370233 Mon Sep 17 00:00:00 2001
From: Tim Chase <tim@chase2k.com>
Date: Fri, 25 Mar 2016 15:35:15 -0400
Subject: [PATCH] 6841 Undirty freed spill blocks Reviewed by: Brian Behlendorf
 <behlendorf1@llnl.gov> Reviewed by: Dan McDonald <danmcd@omniti.com> Reviewed
 by: Matthew Ahrens <mahrens@delphix.com> Approved by: Robert Mustacchi
 <rm@joyent.com>

---
 module/zfs/dbuf.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/module/zfs/dbuf.c b/module/zfs/dbuf.c
index 9b462fd..db9dd5b 100644
--- a/module/zfs/dbuf.c
+++ b/module/zfs/dbuf.c
@@ -908,8 +908,10 @@ dbuf_free_range(dnode_t *dn, uint64_t start_blkid, uint64_t end_blkid,
 	dmu_buf_impl_t *db, *db_next;
 	uint64_t txg = tx->tx_txg;
 	avl_index_t where;
+	boolean_t freespill =
+	    (start_blkid == DMU_SPILL_BLKID || end_blkid == DMU_SPILL_BLKID);
 
-	if (end_blkid > dn->dn_maxblkid && (end_blkid != DMU_SPILL_BLKID))
+	if (end_blkid > dn->dn_maxblkid && !freespill)
 		end_blkid = dn->dn_maxblkid;
 	dprintf_dnode(dn, "start=%llu end=%llu\n", start_blkid, end_blkid);
 
@@ -918,7 +920,7 @@ dbuf_free_range(dnode_t *dn, uint64_t start_blkid, uint64_t end_blkid,
 	db_search.db_state = DB_SEARCH;
 
 	mutex_enter(&dn->dn_dbufs_mtx);
-	if (start_blkid >= dn->dn_unlisted_l0_blkid) {
+	if (start_blkid >= dn->dn_unlisted_l0_blkid && !freespill) {
 		/* There can't be any dbufs in this range; no need to search. */
 #ifdef DEBUG
 		db = avl_find(&dn->dn_dbufs, &db_search, &where);
-- 
1.8.3.1

